finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,719 @@
+use crate::{
+    new_network::{
+        connection_manager::{Authentication, Multiaddr, SessionHandler},
+        DataCommand, PeerId, Protocol,
+    },
+    NodeCount, NodeIndex, SessionId,
+};
+use codec::{Decode, Encode};
+use std::{
+    collections::{HashMap, HashSet},
+    time::{Duration, Instant},
+};
+
+/// How many nodes we should query about unknown authorities in one go.
+const NODES_TO_QUERY: usize = 2;
+
+/// Messages used for discovery and authentication.
+#[derive(Clone, Debug, Encode, Decode)]
+pub enum DiscoveryMessage {
+    AuthenticationBroadcast(Authentication),
+    // Requests always contain own authentication, to avoid asymmetric trust.
+    Request(Vec<NodeIndex>, Authentication),
+    // Always assumed to contain only authentications for one session.
+    // Only authentications from the same session as the first present are guaranteed to be
+    // processed.
+    Authentications(Vec<Authentication>),
+}
+
+impl DiscoveryMessage {
+    pub fn session_id(&self) -> SessionId {
+        use DiscoveryMessage::*;
+        match self {
+            AuthenticationBroadcast((auth_data, _)) => auth_data.session(),
+            Request(_, (auth_data, _)) => auth_data.session(),
+            Authentications(auths) => match auths.get(0) {
+                Some((auth_data, _)) => auth_data.session(),
+                None => SessionId(0), // Broken message anyway, value doesn't matter.
```

Comment:
Don't we want to log it somehow in any case?

Reply:
Good point, although I'll probably do this outside of this method (unwritten code as of now <_<"). Here I only want to return a session id without worrying about this corner case.